server/src/storage/data_transfer.rs

```diff
@@ -0,0 +1,399 @@
+    /// and add it to the total size of the current local files under the key.
+    pub fn add_compressed_file(&mut self, key: &u64, file_path: &Path) -> Result<(), ParquetError> {
+        let file_size = file_path.metadata()?.len() as usize;
+        *self.compressed_files.entry(*key).or_insert(0) += file_size;
+
+        // If the combined size of the files is larger than the batch size, transfer the data to the blob store.
+        if self.compressed_files.get(key).unwrap() >= &self.transfer_batch_size_in_bytes {
+            self.transfer_data(key)?;
+        }
+
+        Ok(())
+    }
+
+    /// Transfer the data corresponding to `key` to the blob store. Once successfully transferred,
+    /// the data is deleted from local storage. Return [`Ok`] if the files were transferred
+    /// successfully, otherwise [`ParquetError`].
+    fn transfer_data(&mut self, key: &u64) -> Result<(), ParquetError> {
+        let _span = debug_span!("transfer_data", key = key).entered();
+
+        // Read all files that correspond to the key.
+        let object_metas = self.runtime.block_on(async {
+            let path = format!("{}/compressed", key).into();
+            let list_stream = self.data_folder_object_store
+                .list(Some(&path))
+                .await?;
+
+            Ok(list_stream.filter_map(|maybe_meta| async {
+                maybe_meta.ok()
+            }).collect::<Vec<_>>().await.into_iter())
+        }).map_err(|error: object_store::Error| ParquetError::General(error.to_string()))?;
+
+        debug!("Transferring {} compressed files.", object_metas.len());
+
+        // Combine the Apache Parquet files into a single RecordBatch.
+        let record_batches = object_metas.clone().filter_map(|meta| {
+            let path = self.data_folder_path.to_string_lossy();
+            let file_path = PathBuf::from(format!("{}/{}", path, meta.location));
+
+            StorageEngine::read_entire_apache_parquet_file(file_path.as_path()).ok()
+        }).collect::<Vec<RecordBatch>>();
+
+        let schema = record_batches[0].schema();
+        let combined = compute::concat_batches(&schema, &record_batches)?;
+
+        debug!("Combined compressed files into single record batch with {} rows.", combined.num_rows());
+
+        // Write the combined RecordBatch to a bytes buffer.
+        let mut buf = vec![].writer();
+        let mut arrow_writer = storage::create_apache_arrow_writer(&mut buf, schema)?;
+        arrow_writer.write(&combined)?;
+        arrow_writer.close()?;
+
+        self.runtime.block_on(async {
+            // Transfer the read data to the blob store.
+            let file_name = storage::create_time_range_file_name(&combined);
+            let path = format!("{}/compressed/{}", key, file_name).into();
+            self.target_object_store.put(&path, Bytes::from(buf.into_inner())).await?;
+
+            // Delete the transferred files from local storage.
```

Comment:
There seems to be a possible issue if the system is terminated right after a transfer have finished successfully but before the transferred files have been deleted from local storage. In this case it seems like the files will be transferred again when the system is restarted. Also it is unclear if it always is safe to delete the files after they have been transferred, e.g., if the query engine is currently using them. Is this the intended behavior or should a task be created to resolve these issues before/when the `DataTransfer` is integrated? One way to manage when files can be used and deleted could be to use a epoch-based file-manager similar to [Heracles](http://vldb.org/pvldb/vol14/p1080-wang.pdf)' epoch-based memory-manager described in Section 3.2. However, I think this is outside the scope of this PR.

Reply:
I added a task to the data transfer component work item that relates to this issue. I agree that it is currently problematic. I have added a link to this comment to the item description. 

I think that task will probably be turned into a separate work item at some point and maybe combined with the [work item](https://dev.azure.com/ModelarData/ModelarDB%20Implementation/_workitems/edit/396) to create better in-memory handling of compressed files to avoid collisions.

Reply:
Shouldn't this also be mentioned in the list of TODOs in the beginning of the file?

Reply:
Good point, I have added it to the todos.